Add eslintrc for CodeFactor.io

Holzhaus opened the pull request:
This adds a custom `.eslint.json` that's better suited for linting controller scripts. It will automatically be picked up by Codefactor.io and follow [our JS coding conventions](https://www.mixxx.org/wiki/doku.php/contributing_mappings#coding_conventions_for_javascript) more closely than the default config.

You can also use it locally, i.e.:
```bash
# Lint single file
$ eslint res/controllers/Roland_DJ-505-scripts.js
# Lint all *.js files in the repository
$ eslint .
```

Holzhaus committed:
.eslintrc.json: Add codefactor.io's default eslint config

This config file is used by codefactor.io and can be used to override
some of the default settings. The default config has been obtained from:
https://github.com/codefactor-io/default-configs/blob/master/.eslintrc.json

Holzhaus committed:
.eslintrc.json: Disable no-unused-vars for function arguments

Previously, this setting led to a lot of false positives in out
controller scripts, e.g.:

    'group' is defined but never used.
    Found in res\controllers\KANE_QuNeo_scripts.js:1120
    KANE_QuNeo.scrollPlaylist = function (channel, control, value, status, group) {

Holzhaus committed:
.eslintrc.json: Set up env for QScript controller mappings

Holzhaus committed:
.eslintrc.json: Configure eslint to use honor Mixxx coding conventions

JS coding conventions can be found at:
https://www.mixxx.org/wiki/doku.php/contributing_mappings#coding_conventions_for_javascript

Holzhaus committed:
.eslintrc.json: Add '_' and 'color' to list of globals

Holzhaus committed:
.eslintrc.json: Re-enable some recommended linter rules

Holzhaus committed:
.eslintrc.json: Reduce global variable scope to certain source dirs

Holzhaus committed:
res/controllers/novation-launchpad: Remove .eslintrc.yaml

The .eslintrc.yaml for novation-launchpad has issues with recent
versions of eslint and requires installation of custom npm packages.

    $ eslint .

    Oops! Something went wrong! :(

    ESLint: 6.8.0.

    ESLint couldn't find the config "standard" to extend from. Please check that the name of the config is correct.

    The config "standard" was referenced from the config file in "/data/jan/Projects/mixxx/res/controllers/novation-launchpad/.eslintrc.yaml".

    If you still have problems, please stop by https://gitter.im/eslint/eslint to chat with the team.

Holzhaus committed:
.eslintignore: Ignore the res/controllers/lodash.mixxx.js script

Holzhaus commented:
By the way, I also auto-fixed a lot of the new issues, but I didn't know if such a mass-change of controller scripts is desired. I pushed it to another branch: https://github.com/mixxxdj/mixxx/commit/82606870a84c6fdb9ae134200382bed46de5f70e

Holzhaus commented:
I removed all style fixes in `*.js*` files and rebased on 2.2.

@Be-ing I'd like to merge this ASAP. The improved linter settings should make reviewing controller script PRs much easier.

uklotzde commented:
I like the Rust naming convention of prepending an underscore to unused variables and args:

```diff
-"no-unused-vars": ["error", { "vars": "all", "args": "none" }],
+"no-unused-vars": ["error", { "vars": "all", "args": "all" }, { "argsIgnorePattern": "^_" }],
```

Implicitly ignoring all unused args may occlude obsolete and redundant code.

uklotzde commented:
...maybe we could then stay with the default settings for vars ("all") and args ("after-used"):

```
"no-unused-vars": ["error", { "argsIgnorePattern": "^_" }],
```

Holzhaus committed:
.eslintrc.json: Only ignore unused args if they start with an underscore

uklotzde commented on .eslintrc.json:
```diff
@@ -5,7 +5,7 @@
   },
   "rules": {
-    "no-unused-vars": ["error", { "vars": "all", "args": "none" }],
+    "no-unused-vars": ["error", { "argsIgnorePattern": "^_" }],
```
:+1: With this convention we can declare all available callback function parameters in controller scripts, even if we don't need them.

Holzhaus committed:
.eslintrc.json: Re-enabled remaining recommended eslint rules

These shouldn't cause a lot of linter issues in our existing codebase
(if any). There's no reason why they shouldn't be enabled IMHO.

Holzhaus committed:
.eslintrc.json: Enable require-atomic-updates rule

Although Promises aren't support by Qt Script anyway, some scripts in
the res/controllers directory are using Promises anyway (i.e. the Novation
Launchpad scripts). So in any case it doesn't hurt to enable this rule.

Holzhaus committed:
.eslintrc.json: Order rules alphabetically

Holzhaus committed:
.eslintrc.json: Reduce camelcase severity to "warn"

Holzhaus committed:
.eslintrc.json: Enforce consistent "dot-location" style

Holzhaus committed:
.eslintrc.json: Add even more code style rules

Holzhaus committed:
.eslintrc.json: Allow opening/closing braces for block on same line

Holzhaus commented:
I added some more code style enforcements regarding braces and spacing. Yes, this raises a lot of issues with our existing codebase, but it will make sure that all new controller scripts follow a consistent code style. CodeFactor will only complain about new issues introduced by a PR, not those that are already present, so this shouldn't be a problem.

@uklotzde @Be-ing Please review the rules again. I think all of them make sense -  let me know if you agree with them.

Holzhaus committed:
.eslintrc.json: Add more rules regarding key spacing and trailing spaces

uklotzde commented:
LGTM. Even VS Code picked up the rules.

Holzhaus commented:
Since this is just a linter config improvement that will make controller mapping reviews much easier and can easily be improved further later on, I'll merge this now.

Holzhaus committed:
.eslintrc.json: Disallow unneeded ternary ops and spaces before func parens

uklotzde commented:
:+1: I was unsure if you were waiting for @Be-ing's response and didn't merge it myself only for this reason.
